File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include<iostream>
#include <vector>
#include <queue>

using namespace std;



int solution(vector<vector<int>> triangle) {
    int answer = 0;
    int size = triangle.size();
   
    queue<int>q;
    
    for (int i = 0; i < triangle[size-1].size(); i++) {
        q.push(triangle[size-1][i]);
    }

    for (int i = size-2; i >=0; i--) {
        for (int j = 0; j < triangle[i].size(); j++) {
            int num1 = q.front(); q.pop();
            int num2 = q.front();
            if (num1 > num2) {
                q.push(num1 + triangle[i][j]);
            }
            else {
                q.push(num2 + triangle[i][j]);
            }
        }
        q.pop();
    }

    answer = q.front(); q.pop();
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    int [][] dp;
    
    public int solution(int[][] triangle) {
        int answer = 0;
        int n = triangle.length;
        dp = new int [n][n];
        
        // 초기 설정 세팅
        for(int i=0;i<n;i++){
            dp[n-1][i] = triangle[n-1][i];
        }
        
        for(int i=n-2;i>=0;i--){ // i는 층
            for(int j=0;j<triangle[i].length;j++){ // 각 층마다
                dp[i][j] = (int)Math.max(dp[i+1][j], dp[i+1][j+1])+triangle[i][j];
            }
        }
    
        
        answer = dp[0][0];
        
        return answer;
    }
}